Improve Installed Operators list page

rebeccaalpert opened the pull request:
Fixes 98% of https://issues.redhat.com/browse/CONSOLE-2130.

## Changes
- I removed the "Deployment" column and added a "Last Updated" column.
- If "All Projects" is selected in the project selector, there is 1 row per installation of an operator/subscription:
    - If Operator X is installed globally, there is 1 row
    - If Operator X is installed in one namespace, there is 1 row
    - If Operator X is installed twice in Namespace A and Namespace B, there are 2 rows, a row per install.
- I added an "Operator Namespace" column indicating where the operator itself is installed (present on all projects view).
    - The column includes a "namespace-name" which is linked to the namespace
    - If the operator is installed globally, the muted text below "namespace-name" is "All Namespaces"
- I added a "Managed Namespace(s)" column indicating the namespace(s) that the operator is watching.
    - If the operator is global, it lists "All Namepaces" in muted text
    - If the operator is not global, it lists the namespace(s) that it is watching
    - I added a popover to list out namespaces if the list is 2+ to ensure the row doesn't blow up
- I changed the "Namespace" field on the Operator Details page to "Managed Namespace"
- I changed the "Namespace" field on the Operator Subscriptions page to "Operator Namespace"

The PatternFly React table does not currently support popovers in sortable headers, so that will be done as a follow-on. (I created this issue to track this: https://issues.redhat.com/browse/CONSOLE-2168.)

## Screenshots
Before:
<img width="995" alt="Screen Shot 2020-03-30 at 12 12 15 PM" src="https://user-images.githubusercontent.com/7014965/77936474-f630e000-7280-11ea-91da-721538eb758c.png">

After (all projects):
<img width="1034" alt="Screen Shot 2020-03-30 at 5 31 04 PM" src="https://user-images.githubusercontent.com/7014965/77963959-47a29480-72ac-11ea-989c-b316da9614e7.png">

After (single project):
<img width="1034" alt="Screen Shot 2020-03-30 at 5 30 04 PM" src="https://user-images.githubusercontent.com/7014965/77963915-348fc480-72ac-11ea-9b2c-b93d05799792.png">

Multinamespace:
<img width="964" alt="Screen Shot 2020-03-30 at 12 25 53 PM" src="https://user-images.githubusercontent.com/7014965/77936982-a999d480-7281-11ea-9588-c395d0aec0d4.png">

All namespaces:
<img width="999" alt="Screen Shot 2020-03-30 at 2 41 04 PM" src="https://user-images.githubusercontent.com/7014965/77949202-89bfdc00-7294-11ea-8abf-04a77e608bb1.png">

Subscription:
![77947237-a0b0ff00-7291-11ea-96f3-b124ad29f7bb](https://user-images.githubusercontent.com/7014965/78054245-c00e6180-734f-11ea-8086-bd9984a87762.png)

CSV Details Page:
<img width="756" alt="Screen Shot 2020-03-30 at 5 28 27 PM" src="https://user-images.githubusercontent.com/7014965/77963801-014d3580-72ac-11ea-8d08-1ce8b1fb7537.png">

Subscription Details Page:
<img width="576" alt="Screen Shot 2020-03-31 at 12 22 07 PM" src="https://user-images.githubusercontent.com/7014965/78050515-517ad500-734a-11ea-83e3-45f2c9231959.png">

Heads up @openshift/team-ux-review.

rebeccaalpert commented:
/retest

"Docker build strategy has failed."

rachael-phillips commented:
Hi @rebeccaalpert still taking a look at this, but at first glance we should add The Namespace where the Operator is installed to the Subscription page popover. 

![image](https://user-images.githubusercontent.com/40275806/78059366-b983e800-7357-11ea-8d24-503bcc73f9a8.png)

rebeccaalpert commented:
All set @rachael-phillips!

<img width="566" alt="Screen Shot 2020-03-31 at 2 03 08 PM" src="https://user-images.githubusercontent.com/7014965/78059818-8aba4180-7358-11ea-8767-193c06305493.png">

christianvogt commented on frontend/public/components/utils/details-item.tsx:
```diff
@@ -43,8 +43,15 @@ export const DetailsItem: React.FC<DetailsItemProps> = ({
+  }
+  if (label === 'Operator Namespace' && obj.kind === 'Subscription') {
+    description = `The Namespace where the Operator is installed.\n\n${propertyDescription}`;
+  }
```
Same comment here. We should find a way to provide this information from the parent component.

christianvogt commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -421,8 +511,51 @@ export const ClusterServiceVersionList: React.SFC<ClusterServiceVersionListProps
+          if (source === undefined || source === null) {
+            return null;
+          }
+          if (source.kind !== 'ClusterServiceVersion') {
```
Instead of a string, can you reference the Model kind?

rebeccaalpert replied:
I'll push an updated version after responding to the other feedback; let me know if you want a different approach once you can see it.

christianvogt commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -421,8 +511,51 @@ export const ClusterServiceVersionList: React.SFC<ClusterServiceVersionListProps
+              source.metadata.annotations &&
+              source.metadata.annotations['olm.operatorGroup'] !== undefined &&
+              source.metadata.annotations['olm.operatorGroup'] !== null &&
+              source.metadata.annotations['olm.operatorGroup'] === 'global-operators'
```
```suggestion
              source.metadata.annotations?.['olm.operatorGroup'] === 'global-operators'
```

christianvogt commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -421,8 +511,51 @@ export const ClusterServiceVersionList: React.SFC<ClusterServiceVersionListProps
+  const filteredData =
+    data.length > 0
+      ? data.filter((source) => {
+          if (source === undefined || source === null) {
```
```suggestion
          if (source == null) {
```

christianvogt commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -209,12 +256,19 @@ const ClusterServiceVersionStatus: React.FC<ClusterServiceVersionStatusProps> =
+      'olm.operatorGroup': olmOperatorGroup = '',
+      'olm.targetNamespaces': olmTargetNamespaces = '',
+      'olm.operatorNamespace': olmOperatorNamespace = '',
+    } = _.get(obj, 'metadata.annotations') || {};
```
Prefer optional chaining over lodash `get`.
```suggestion
    } = obj.metadata?.annotations || {};
```

christianvogt commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.spec.tsx:
```diff
@@ -116,24 +115,16 @@ describe(ClusterServiceVersionTableRow.displayName, () => {
-    expect(link.props().name).toEqual(testClusterServiceVersion.metadata.namespace);
+  it('renders column for managedNamespace', () => {
+    const col = wrapper.find(TableRow).childAt(1);
+    const managedNamespace = col.childAt(0).find('span.text-muted');
```
This is a very vague selector. Can you use something more stable?

christianvogt commented on frontend/public/components/factory/table.tsx:
```diff
@@ -104,12 +105,38 @@ const sorts = {
+    if (
+      !_.isUndefined(olmOperatorGroup) &&
+      olmOperatorGroup !== null &&
+      olmOperatorGroup === 'global-operators'
```
```suggestion
      olmOperatorGroup === 'global-operators'
```

christianvogt commented on frontend/public/components/factory/table.tsx:
```diff
@@ -104,12 +105,38 @@ const sorts = {
+    ) {
+      olmManagedNamespace = 'All Namespaces';
+    } else {
+      const namespaces = _.get(obj, 'metadata.annotations["olm.targetNamespaces"]');
```
```suggestion
      const namespaces = obj.metadata?.annotations?.['olm.targetNamespaces'];
```

christianvogt commented on frontend/public/components/factory/table.tsx:
```diff
@@ -104,12 +105,38 @@ const sorts = {
+      const namespacesList = _.isUndefined(namespaces) ? [] : namespaces.split(',');
+      namespacesList.length > 1
+        ? (olmManagedNamespace = `{namespacesList.length} Namespaces`)
+        : (olmManagedNamespace = namespacesList);
```
Missing template `$` ?
```suggestion
      olmManagedNamespace = namespacesList.length > 1 ?`${namespacesList.length} Namespaces` : namespacesList;
```

christianvogt commented on frontend/public/components/factory/table.tsx:
```diff
@@ -104,12 +105,38 @@ const sorts = {
   serviceCatalogStatus,
   jobCompletions: (job) => getJobTypeAndCompletions(job).completions,
   jobType: (job) => getJobTypeAndCompletions(job).type,
+  managedNamespace: (obj) => {
```
Is there any way to contribute these specific sorts from OLM through extensions? if not we should consider adding something.

rebeccaalpert replied:
It looks like operand in operator-lifecycle-manager package isn't listing its sort here and keeps it within the package instead; I can go with that approach if that works.

christianvogt commented on frontend/public/components/factory/table.tsx:
```diff
@@ -104,12 +105,38 @@ const sorts = {
   },
   numReplicas: (resource) => _.toInteger(_.get(resource, 'status.replicas')),
+  operatorNamespace: (obj) => {
+    const olmOperatorNamespace = _.get(obj, 'metadata.annotations["olm.operatorNamespace"]');
```
```suggestion
    const olmOperatorNamespace = obj.metadata?.annotations?.['olm.operatorNamespace'];
```

christianvogt commented on frontend/public/components/factory/table.tsx:
```diff
@@ -104,12 +105,38 @@ const sorts = {
+  operatorNamespace: (obj) => {
+    const olmOperatorNamespace = _.get(obj, 'metadata.annotations["olm.operatorNamespace"]');
+    const namespace = getNamespace(obj);
+    return !_.isUndefined(olmOperatorNamespace) ? olmOperatorNamespace : namespace;
```
```suggestion
    return olmOperatorNamespace ?? getNamespace(obj);
```

christianvogt commented on frontend/public/components/factory/table.tsx:
```diff
@@ -454,11 +481,18 @@ export const Table = connect<
 
     componentDidMount() {
-      const { columns } = this.state;
+      const componentProps: any = _.pick(this.props, [
```
We should avoid `any` type.

rebeccaalpert replied:
I didn't write this one and I'm honestly not sure how to better define the type given that it changes so much in action.

christianvogt commented on frontend/public/components/factory/table.tsx:
```diff
@@ -500,7 +534,15 @@ export const Table = connect<
     _onSort(event, index, direction) {
       event.preventDefault();
-      const sortColumn = this.state.columns[index - this._columnShift];
+      const componentProps: any = _.pick(this.props, [
```
We should avoid `any` type.

spadgett commented:
> Hi @rebeccaalpert still taking a look at this, but at first glance we should add The Namespace where the Operator is installed to the Subscription page popover.
> 
> ![image](https://user-images.githubusercontent.com/40275806/78059366-b983e800-7357-11ea-8d24-503bcc73f9a8.png)

I don't think we should modify the popover.

These descriptions come entirely from the API documentation today. I think it's a bad idea to diverge from our API doc and special case certain fields in console with special documentation. What you see in the popover should match `oc explain`. The reason for this approach is to let the owners of the resources provide accurate documentation and to make sure it's always up-to-date (since API doc will be updated when APIs change).

spadgett commented on frontend/public/components/utils/details-page.tsx:
```diff
@@ -65,11 +65,19 @@ export const ResourceSummary: React.SFC<ResourceSummaryProps> = ({
     namespace: metadata.namespace,
   });
 
+  let label = 'Namespace';
+  if (kind === 'Subscription') {
+    label = 'Operator Namespace';
+  }
+  if (kind === 'ClusterServiceVersion') {
+    label = 'Managed Namespace';
+  }
```
I agree with @christianvogt that this component should be generic and not know about these kinds.

From a UX standpoint, I also think this change is confusing as well. Calling it something other than `Namespace` makes me think the resource is in a different namespace and is just managing this namespace, which is not the case. It makes me think I can change the managed namespace, which I can't.

This component is here to ensure consistency across pages, and every other page calls this `Namespace`. `Namespace` also matches what the CLI and YAML show. It's hard to change this label because it *should* be hard to change this label. Changing it creates an inconsistency.

rachael-phillips replied:
@alimobrem @tlwu2013 @dmesser what are your thoughts on this? Curious to hear what you think, as well.

itsptk replied:
I see the value in indicating to the user what the actual labels are so that they can map them to the CLI/YAML, and also pulling the contents of the popover from API descriptions to keep them current. I do think though, that doesn't leave much room in these details screens to inject any additional UX niceness (further describing a namespace as "operator" or "managed" for example.)

Perhaps we could have some common paradigm where these labels are allowed to be changed in the UI (some sort of 'display name') and if that is present, add the actual resource label name is added to the popover contents for reference.

"Appears as `Namespace` in the `ClusterServiceVersion` YAML."

Alternatively, maybe these popovers could be divided into two sections, an optional 'ui-specified' description that appears first, followed by the API description.

It would seem unfortunate to try to further describe these two kinds of namespaces in the list view's column headers and not have any indication of which are which in the resource details.

itsptk commented:
> Subscription:
> ![77947237-a0b0ff00-7291-11ea-96f3-b124ad29f7bb](https://user-images.githubusercontent.com/7014965/78054245-c00e6180-734f-11ea-8086-bd9984a87762.png)

@rebeccaalpert I was wondering what view this screen was conveying? Is this when there is no subscription associated with the CSV?

rebeccaalpert commented:
Hey @itsptk - Subscriptions show up in the Installed Operators view while the operator is installing or if it fails to install. (The row disappears once the operator is successfully installed and is replaced with the operator's details). This screen illustrates how subscriptions handle the new columns like "Managed Namespace."

rebeccaalpert commented:
I updated the code to reflect changes discussed on Slack. I addressed code feedback (except for the use of any in the table). I wasn't able to get any information on better types for the props involved. Happy to change it if anyone can provide information.

All projects:
<img width="1033" alt="Screen Shot 2020-04-03 at 1 51 05 PM" src="https://user-images.githubusercontent.com/7014965/78399005-1da9e480-75c2-11ea-802e-6868ea92e1ab.png">

Single project:
<img width="1029" alt="Screen Shot 2020-04-03 at 1 52 39 PM" src="https://user-images.githubusercontent.com/7014965/78398984-108cf580-75c2-11ea-9c50-84b32d7586d1.png">

All Namespaces:
<img width="769" alt="Screen Shot 2020-04-03 at 10 57 17 AM" src="https://user-images.githubusercontent.com/7014965/78398904-eaffec00-75c1-11ea-8650-6b278cda115b.png">

Single Namespace:
<img width="772" alt="Screen Shot 2020-04-03 at 1 53 06 PM" src="https://user-images.githubusercontent.com/7014965/78398926-f226fa00-75c1-11ea-8881-27c0b1c6e459.png">

Managed Namespaces Popover:
<img width="446" alt="Screen Shot 2020-04-03 at 1 55 10 PM" src="https://user-images.githubusercontent.com/7014965/78398890-e5a2a180-75c1-11ea-94f3-92b38b342618.png">

rebeccaalpert commented:
/retest

rebeccaalpert commented:
```
error: error creating buildah builder: Error committing the finished image: error adding layer with blob "sha256:550d2fcafdc527da6b2d6f9407757935dc328dcc01f5be2d87d3bf58daf3d80d": open /var/run/containers/storage/overlay-layers/.tmp-mountpoints.json973470951: no space left on device
66
2020/04/06 14:27:58 No custom metadata found and prow metadata already exists. Not updating the metadata.
67
2020/04/06 14:27:59 Ran for 5m44s
68
2020/04/06 14:27:59 Submitted failure event to sentry (id=684601af0a694bdda8ed73e57c43969b)
69
error: could not run steps: step console failed: could not wait for build: the build console failed after 41s with reason GenericBuildFailed: Generic Build failure - check logs for details.
```

/retest

rebeccaalpert commented:
```
2020/04/06 14:57:23 Building test-bin
2020/04/06 14:57:23 Build test-bin failed, printing logs:
unable to retrieve container logs for cri-o://a4459410eb9ba000476c73c8311b85696dfafb709f9fdf19195e19bedbf5a8b0unable to retrieve container logs for cri-o://5c2407219f975b2cecc7fe8ee1a1f6ab43afaf2cfa8b377373e052f346d88e902020/04/06 14:57:23 No custom metadata found and prow metadata already exists. Not updating the metadata.
2020/04/06 14:57:24 Ran for 1s
2020/04/06 14:57:24 Submitted failure event to sentry (id=fd7ccacb10904272a75fb131d174dd7d)
error: could not run steps: step test-bin failed: could not wait for build: the build test-bin failed with reason DockerBuildFailed: Dockerfile build strategy has failed.
error https://registry.yarnpkg.com/@fortawesome/fontawesom...s/@fortawesome/fontawesome-free/svgs/brands/app-store.svg'"
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
subprocess exited with status 1
subprocess exited with status 1
error: build error: error building at STEP "RUN /bin/bash -c set -o errexit; umask 0002; ./build.sh": exit status 1
```

rebeccaalpert commented:
/retest

rebeccaalpert commented:
```
2020/04/06 15:12:56 Building src
2020/04/06 15:12:56 Build src already succeeded in 3m7s
2020/04/06 15:12:56 Building test-bin
2020/04/06 15:12:56 Build test-bin failed, printing logs:
unable to retrieve container logs for cri-o://a4459410eb9ba000476c73c8311b85696dfafb709f9fdf19195e19bedbf5a8b0unable to retrieve container logs for cri-o://5c2407219f975b2cecc7fe8ee1a1f6ab43afaf2cfa8b377373e052f346d88e902020/04/06 15:12:56 No custom metadata found and prow metadata already exists. Not updating the metadata.
2020/04/06 15:12:57 Ran for 2s
2020/04/06 15:12:57 Submitted failure event to sentry (id=19ddd93b36554eb9888822c6890e919e)
error: could not run steps: step test-bin failed: could not wait for build: the build test-bin failed with reason DockerBuildFailed: Dockerfile build strategy has failed.
error https://registry.yarnpkg.com/@fortawesome/fontawesom...s/@fortawesome/fontawesome-free/svgs/brands/app-store.svg'"
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
subprocess exited with status 1
subprocess exited with status 1
error: build error: error building at STEP "RUN /bin/bash -c set -o errexit; umask 0002; ./build.sh": exit status 1
```

rebeccaalpert commented:
/retest

spadgett commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -422,8 +512,72 @@ export const ClusterServiceVersionList: React.SFC<ClusterServiceVersionListProps
+        managedNamespace: (obj) => {
+          const olmOperatorGroup = obj.metadata?.annotations?.['olm.operatorGroup'];
+          let olmManagedNamespace;
+          if (olmOperatorGroup === 'global-operators') {
```
I don't think this is right for native k8s. I believe it's more correct to just return `All Namespaces` when the annotation is missing from the CSV.

spadgett commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -440,6 +594,28 @@ export const ClusterServiceVersionList: React.SFC<ClusterServiceVersionListProps
       NoDataEmptyMsg={NoDataEmptyMsg}
       virtualize
+      customSorts={{
+        managedNamespace: (obj) => {
```
Don't duplicate this logic... create a function outside of the component that you reference here

benjaminapetersen replied:
+1 great time to deduplicate a couple of things.

spadgett commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -686,7 +864,25 @@ export const ClusterServiceVersionDetails: React.SFC<ClusterServiceVersionDetail
+                  description: 'Operands in this Namespace are managed by the Operator.',
+                  path: 'metadata.namespace',
+                  children:
+                    olmOperatorGroup === 'global-operators' ? (
```
Again, this logic isn't completely correct. Check if there is no `olm.targetNamespaces` annotation.

spadgett replied:
Is there a reason we don't handle the case where there are multiple namespaces here like in the list view?

rebeccaalpert replied:
This was @tlwu2013's rationale:

> Details view no need to display multi-NS rare case, reason being:
this is reserved for list view “all-projects” case because only admin users can view all-projects  so they have access to any projects in a cluster
details view is single NS, users may be non-admin role who only have access to their own projects
       --> therefore, the exact same logic as list view “single project” case  (see login pinned in this channel)
       --> i.e. for global Operators, show “All Namespaces” in Managed Namespaces field
                   for non-global Operators, show “[current viewing namespace]” in Managed Namespaces field
hope this helps

spadgett commented on frontend/public/components/utils/details-item.tsx:
```diff
@@ -36,14 +36,16 @@ export const DetailsItem: React.FC<DetailsItemProps> = ({
   const reference: K8sResourceKindReference = referenceFor(obj);
   const model: K8sKind = modelFor(reference);
-  const description: string = getPropertyDescription(model, path);
+  const description: string = customDescription ?? getPropertyDescription(model, path);
```
This doesn't seem like the right way to handle this. This component is specifically for getting API doc and has a lot of logic you don't need. I would just create the custom popover in the one place you need it. It's pretty trivial to add.

spadgett commented on frontend/public/components/utils/details-page.tsx:
```diff
@@ -78,6 +79,16 @@ export const ResourceSummary: React.SFC<ResourceSummaryProps> = ({
         </DetailsItem>
       )}
+      {customResource && (
+        <DetailsItem
```
Don't use `DetailsItem` here. Just create the popover yourself.

spadgett commented on frontend/public/components/utils/details-page.tsx:
```diff
@@ -78,6 +79,16 @@ export const ResourceSummary: React.SFC<ResourceSummaryProps> = ({
           />
         </DetailsItem>
       )}
+      {customResource && (
```
I wouldn't add a `customResource` to the generic summary component. Just make the additional details items children of the component like we do in other places.

If we *really* don't like the ordering it gives, I would just not use the `ResourceSummary` on the CSV page.

spadgett commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -422,8 +512,72 @@ export const ClusterServiceVersionList: React.SFC<ClusterServiceVersionListProps
+          }
+          if (referenceForModel(ClusterServiceVersionModel) === kind) {
+            if (source.metadata.annotations?.['olm.operatorGroup'] === 'global-operators') {
+              return source.metadata.namespace === 'openshift-operators' ? source : null;
```
This logic is wrong for native k8s. The namespace could be something else. Ideally you'd find the right operator group and check its namespace, although I'm not sure that's possible for normal users.

We might have to circle back on this and decide what to do.

spadgett replied:
@rebeccaalpert I would filter out any CSV with status "Copied." This is simpler and should work in all environments.

rebeccaalpert replied:
Thanks, will do.

spadgett commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -339,26 +419,35 @@ const SubscriptionTableRow: React.FC<SubscriptionTableRowProps> = ({
-        <span className="text-muted">None</span>
+      {/* Last Updated */}
+      <TableData className={tableColumnClasses[4]}>
+        {obj.status && fromNow(obj.status.lastUpdated)}
```
You need a fallback value like `-` if `lastUpdated` isn't set in the resource.

rachael-phillips replied:
Hey @spadgett I've noticed in the UI we often just leave the field blank. For the Last Updated field, as a user if this info was not provided I would expect to see something more like "not available" or as we handle it now, blank. Is there a reason that you would opt for the dash over these?

benjaminapetersen replied:
Whatever our preferred default is, we could likely create a simple component that wraps and displays `value || default`.  Eliminate some review maintenance.

spadgett replied:
Usually we show `-` or occassionally some muted text like `No <blank>`. I can't think of a place where we leave the table cell blank (but I'm sure we do somewhere).

spadgett commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -306,6 +385,7 @@ const SubscriptionTableRow: React.FC<SubscriptionTableRowProps> = ({
   const csvName = _.get(obj, 'spec.name');
   const menuActions = [Kebab.factory.Edit, () => uninstall(obj)];
   const namespace = getNamespace(obj);
+  const ns = UIActions.getActiveNamespace();
```
Don't get the namespace this way. Connect to Redux.

spadgett commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -210,12 +257,19 @@ const ClusterServiceVersionStatus: React.FC<ClusterServiceVersionStatusProps> =
 export const ClusterServiceVersionTableRow = withFallback<ClusterServiceVersionTableRowProps>(
   ({ obj, rowKey, subscription, catalogSourceMissing, index, style }) => {
     const { displayName, provider, version } = _.get(obj, 'spec');
+    const ns = UIActions.getActiveNamespace();
```
Don't get the active namespace this way. Connect to Redux.

benjaminapetersen replied:
Does `UIActions.getActiveNamespace()` have a correct use now, or is it tech debt?

spadgett commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -234,13 +288,43 @@ export const ClusterServiceVersionTableRow = withFallback<ClusterServiceVersionT
-        <TableData className={tableColumnClasses[1]}>
-          <ResourceLink kind="Namespace" title={namespace} name={namespace} />
+        {/* Operator Namespace */}
+        {ns === ALL_NAMESPACES_KEY ? (
```
I'm tempted to say we should just always show the column until we can solve this generally for all resources. Ideally I'd like the logic to handle the namespace column in one place in table.tsx instead of having to duplicate the logic in every list view, and I don't want to hold up the important operator table changes on that. Can we address this as a follow on?

benjaminapetersen commented:
![Screen Shot 2020-04-06 at 2 40 29 PM](https://user-images.githubusercontent.com/280512/78595263-bc7b4e80-7817-11ea-8731-4cb6538325c4.png)

Little css blip.

benjaminapetersen commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -100,28 +110,65 @@ export const ClusterServiceVersionTableHeader = () => {
+  ];
+};
+
+export const ClusterServiceVersionTableHeader = () => {
```
nite: I have a tendency to want to deduplicate rather than copy & paste:

```typescript
const sharedFirst = [{title: 'Name'.....}]
const sharedLast = [{title: 'Status'...},{title: 'Last Updated'...},{title: 'Provided APIs'...},{title: ''...}]
const AllProjectsTableHeader = sharedFirst
  .concat([{title: 'Operator Namespace'}])
  .concat(sharedLast)
const SingleProjectTableHeader = sharedFirst
  .concat([{title: 'Operator Namespace'}, {title: 'Managed Namespace'}])
  .concat(sharedLast)
```
But I wouldn't block on it, I'm betting some prefer the simplicty of just listing them directly.

benjaminapetersen replied:
However, this comment does reveal that `props: { className: tableColumnClasses[0,1,2,3,4......] },` may be broken if that was done.

benjaminapetersen replied:
You may instead need a little helper func to give back either of these arrays, but ensure the `tableColumnClasses` list is correct for either table.

rebeccaalpert replied:
The tableColumnClasses are intentional; we only want to display certain columns at certain breakpoints. I can refactor though.

benjaminapetersen replied:
The first is 1,2,3,4,5,6. the other is 0,2,3,4,5,6.  
If the skip is intentional, then 👍 
Looked like the kind of thing that might be an oops :)

rebeccaalpert replied:
I'm refactoring using some string-based classes; let me know if that's easier to read once you can see it (just working out a few linter kinks).

benjaminapetersen commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -100,28 +110,65 @@ export const ClusterServiceVersionTableHeader = () => {
+      title: 'Managed Namespaces',
+      sortFunc: 'managedNamespace',
+      transforms: [sortable, wrappable],
+      props: { className: tableColumnClasses[2] },
```
I think, due to copy/paste, the `props: { className: tableColumnClasses[2] },` is off by 1?  It jumps from `0` to `2` here, between `Name` and `Managed Namespace`.

rebeccaalpert replied:
This was intentional.

benjaminapetersen commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -422,8 +512,72 @@ export const ClusterServiceVersionList: React.SFC<ClusterServiceVersionListProps
   const NoDataEmptyMsg = () => <MsgBox title="No Operators Found" detail={noDataDetail} />;
 
-  return (
+  // remove duplicate global operator entries
```
Instead of a comment, use a named helper function:

```javascript
const filteredData = removedDuplicateGlobalOperators(data);
```
Comments may be removed later, but it is unlikely someone will rename a function in a way that no longer describes its role.

benjaminapetersen replied:
You may want to consider several helpers, to self-document this block:

```typescript
const filteredData = removedDuplicateGlobalOperators(data);

const removedDuplicateGlobalOperators = (operators) => {
  return operators.length > 0
          // filter really responds to true/false 
          ? operators.filter((source) => {
              const kind = referenceFor(source);
              if(isFoo(kind)) {
                return false 
              }
              if(isBar(kind)) {
                return true
              }
              if(isBaz(kind)) {
                return false;
              }
          })
          : [];
} 
```

benjaminapetersen commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -422,8 +512,72 @@ export const ClusterServiceVersionList: React.SFC<ClusterServiceVersionListProps
+      NoDataEmptyMsg={NoDataEmptyMsg}
+      virtualize
+      customSorts={{
+        managedNamespace: (obj) => {
```
nit: You may want to pull these out as named funcs and reference them, to tidy this up a little bit:

```typescript
      virtualize
      customSorts={{
        managedNamespace: managedNamespaceSort,
        namespace: namespaceSort
      }}

// type these!
const namespaceSort = (obj: T) T2 => {}
const managedNamespaceSort = (obj: T) T2 => {}
```
And then also `type` the sorting funcs.

benjaminapetersen commented on frontend/public/components/factory/table.tsx:
```diff
@@ -382,6 +383,14 @@ type TableOptionProps = {
   UI: any;
 };
 
+type ComponentProps = {
```
Thx, good addition.

benjaminapetersen commented:
Just a few px off on the sorting arrows yet. Not sure if that is new. 

<img width="850" alt="Screen Shot 2020-04-15 at 11 35 46 AM" src="https://user-images.githubusercontent.com/280512/79357074-87ed4e00-7f0d-11ea-805a-fae89e4df945.png">

benjaminapetersen commented:
When I click the `view one more...` link, i get redirected to the `All Instances` tab, rather than the `Details` tab with the list of Provided APIs.  
<img width="374" alt="Screen Shot 2020-04-15 at 11 40 13 AM" src="https://user-images.githubusercontent.com/280512/79357361-e9adb800-7f0d-11ea-847c-5a8c6e3f6215.png">

Goes here:
<img width="588" alt="Screen Shot 2020-04-15 at 11 40 53 AM" src="https://user-images.githubusercontent.com/280512/79357463-0fd35800-7f0e-11ea-93f9-d25e8e92d756.png">

Rather than here:
<img width="714" alt="Screen Shot 2020-04-15 at 11 41 55 AM" src="https://user-images.githubusercontent.com/280512/79357508-211c6480-7f0e-11ea-97f8-ec4df5c958e3.png">

spadgett commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -207,96 +248,119 @@ const ClusterServiceVersionStatus: React.FC<ClusterServiceVersionStatusProps> =
+  const route = resourceObjPath(obj, referenceFor(obj));
+  const uid = getUID(obj);
+  const internalObjects = getInternalObjects(obj);
+  const managedNamespaces = olmTargetNamespaces === '' ? [] : olmTargetNamespaces.split(',');
```
```suggestion
  const managedNamespaces = olmTargetNamespaces?.split(',') || [];
```

spadgett commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -207,96 +248,119 @@ const ClusterServiceVersionStatus: React.FC<ClusterServiceVersionStatusProps> =
+      <TableData className={tableColumnClasses[2]}>
+        {managedNamespaces.length === 0 ? (
+          <span className="text-muted">All Namespaces</span>
+        ) : managedNamespaces.length > 1 ? (
```
Avoid nested ternaries. They tend to be confusing.

spadgett commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -686,7 +847,37 @@ export const ClusterServiceVersionDetails: React.SFC<ClusterServiceVersionDetail
+                  <Popover
+                    headerContent={<div>Managed Namespaces</div>}
+                    bodyContent={
+                      <LinkifyExternal>
```
No reason for `LinkifyExternal` here since we know what the content is (and it has no links).

spadgett commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -686,7 +847,37 @@ export const ClusterServiceVersionDetails: React.SFC<ClusterServiceVersionDetail
+                      kind="Namespace"
+                      name={props.obj.metadata.namespace}
+                      title={props.obj.metadata.uid}
+                      namespace={null}
```
Remove this prop if there is no namespace

spadgett commented on frontend/packages/operator-lifecycle-manager/src/style.scss:
```diff
@@ -54,6 +58,20 @@ $co-affinity-row-margin: 15px;
   display: flex;
 }
 
+.pf-c-table .pf-m-wrap {
+  white-space: normal;
+  vertical-align: top;
+}
+
+.pf-c-table__sort.pf-m-wrap .pf-c-button {
+  display: flex;
+  white-space: normal;
+}
+
+.pf-c-table__sort-indicator {
+  line-height: normal !important;
+}
+
```
This will impact *all* tables. We don't want that. Use BEM classes to scope these changes to the single table we need.

rebeccaalpert replied:
These are PF 4 styles we don't seem to be pulling in (PatternFly officially supports wrappable, sortable columns, but we don't seem to have that support within OpenShift out of the box for some reason). I can scope these, but do you know if there's a reason we don't already have support for this? It would be ideal if I can just drop these.

spadgett replied:
> These are PF4 styles we don't seem to be pulling in (PatternFly officially supports wrappable, sortable columns, but we don't seem to have CSS support within OpenShift out of the box for some reason). I can scope these, but do you know if there's a reason we don't already have support for this? It would be ideal if I could just drop these.

Are we sure we have the right Patternfly version for that? It was added recently.

cc @sg00dwin @rhamilto

rebeccaalpert replied:
It looks like it was added back in August and we're past those versions: https://www.patternfly.org/v4/documentation/react/overview/release-notes#201906-release-notes-2019-08-13

spadgett replied:
> It looks like it was added back in August and we're past those versions: https://www.patternfly.org/v4/documentation/react/overview/release-notes#201906-release-notes-2019-08-13

We need this fix, which is more recent: https://github.com/patternfly/patternfly/pull/2668

rebeccaalpert replied:
It looks like that was part of release 2.65.2 and we're on 2.65.3?

https://www.patternfly.org/v4/documentation/core/overview/release-notes#202002-release-notes-2020-02-18

benjaminapetersen replied:
Seem like we need to do one of the following:
1. bump PF to pick up the fix.
2. scope it to just this instance for now.
3. if not scoped, move it out of `packages/operator-lifecycle-manager`, if it's going to affect all tables until we can bump PF.

spadgett replied:
@rebeccaalpert can you remove these styles from this PR and we'll handle it in #5078 ?

rebeccaalpert replied:
All set!

rebeccaalpert commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -422,11 +502,82 @@ export const ClusterServiceVersionList: React.SFC<ClusterServiceVersionListProps
+      olmManagedNamespace =
+        namespacesList.length > 1 ? `${namespacesList.length} Namespaces` : namespacesList;
+    }
+    if (obj.kind === 'Subscription') {
+      olmManagedNamespace = 'None';
+    }
```
@rachael-phillips and @tlwu2013 suggested using 'None' for the Subscriptions row data for Managed Namespace. This is keeping it consistent to that value. I can refactor the function though.

spadgett replied:
I guess I'm thinking about the case where `None` will flicker and then be replaced when the operator is first installed (since the Subscription will exist before the CSV).

@tlwu2013 Are we trying to handle the case where the CSV is deleted, but the Subscription still exists?

rebeccaalpert commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -422,11 +502,82 @@ export const ClusterServiceVersionList: React.SFC<ClusterServiceVersionListProps
+      : [];
+  };
+
+  const filteredData = removedDuplicateGlobalOperators(data);
```
We are, but I'll refactor to make this clearer.

rebeccaalpert commented:
I pushed the PR feedback I can implement right now.

spadgett commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -79,52 +88,83 @@ import { operatorGroupFor, operatorNamespaceFor } from './operator-group';
+  title: 'Name',
+  sortField: 'metadata.name',
+  transforms: [sortable],
+  props: { className: '' },
```
You should use the values from `tableColumnClasses` or remove it. Maybe it would be better to have constants for each (since the index will be different depending on whether all namespaces is selected).

```ts
const nameColumnClass = '';
```

Then...

```suggestion
  props: { className: nameColumnClass },
```

etc.

spadgett commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -79,52 +88,83 @@ import { operatorGroupFor, operatorNamespaceFor } from './operator-group';
+  props: { className: '' },
+};
+
+const managedNamespaceHeader: Header = {
```
```suggestion
const managedNamespacesHeader: Header = {
```

spadgett commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -207,96 +247,129 @@ const ClusterServiceVersionStatus: React.FC<ClusterServiceVersionStatusProps> =
+  const uid = getUID(obj);
+  const internalObjects = getInternalObjects(obj);
+  const managedNamespaces = olmTargetNamespaces?.split(',') || [];
+  const managedNamespaceData = () => {
```
This would be better as a separate React component instead of a function that returns markup.

spadgett commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -422,11 +511,82 @@ export const ClusterServiceVersionList: React.SFC<ClusterServiceVersionListProps
+      if (isSubscription(kind)) {
+        return true;
+      }
+      if (isCopiedCSV(source, kind)) {
+        return false;
+      }
+      return true;
```
```suggestion
      return !isCopiedCSV(source, kind);
```

spadgett commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -686,7 +856,34 @@ export const ClusterServiceVersionDetails: React.SFC<ClusterServiceVersionDetail
+                  <Popover
+                    headerContent={<div>Managed Namespaces</div>}
+                    bodyContent={
+                      <div className="co-pre-line">
```
`co-pre-line` is not needed here either

rebeccaalpert commented:
Addressed round of PR feedback (refactored into component, renamed column classes, etc.)

benjaminapetersen commented:
/approve

👍 to the approach, getting closer, just a couple items left for the `lgtm`.

rebeccaalpert commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -79,52 +88,82 @@ import { operatorGroupFor, operatorNamespaceFor } from './operator-group';
+const statusColumnClass = classNames('pf-m-hidden', 'pf-m-visible-on-lg');
+const lastUpdatedColumnClass = classNames('pf-m-hidden', 'pf-m-visible-on-2xl');
+const providedAPIsColumnClass = classNames('pf-m-hidden', 'pf-m-visible-on-xl');
+const kebabColumnClass = Kebab.columnClass;
```
All set! (lgtm got removed.)

spadgett commented:
/test frontend

spadgett commented:
/hold
for #5100

spadgett commented:
/hold cancel
/retest
